source/common/formatter/substitution_formatter.cc

```diff
@@ -458,6 +460,12 @@ SubstitutionFormatParser::getKnownFormatters() {
          [](const std::string& format, const absl::optional<size_t>& max_length) {
            return FilterStateFormatter::create(format, max_length, false);
```

Comment:
I wonder if it makes more sense to leverage the filter state formatter. The documentation seems promising. See the `%FILTER_STATE(KEY:F):Z%` section on [this page](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage).
```
F is an optional parameter used to indicate which method FilterState uses for serialization.
```
It sounds like base64 can be used here. I feel it would have a better chance to get merged upstream.

Reply:
Hmm you mean adding a new `F` option in addition to `TYPED` and `PLAIN`? The new serialization method couldn't be as simple as `BASE64`, because we don't want to encode the entire proxy protocol filter state value. We could, however, create a new `TLVS` value to do the full formatting. How does that sound to you, @visualzhou ?

Reply:
I see. Do you mind articulating what TYPED and PLAIN would look like for our case?

Reply:
From their docs: 
```
If ‘PLAIN’ is set, the filter state object will be serialized as an unstructured string. If ‘TYPED’ is set or no F provided, the filter state object will be serialized as an JSON string.
```

For our case, this means that the binary TLV value would not be displayed if either the `PLAIN` or `TYPED` method are used. In my testing, [this call](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/source/common/formatter/substitution_formatter.cc#L1447) to `serializeAsString` (the serializing function called when `PLAIN` is used) returns an empty optional, and [this call](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/source/common/formatter/substitution_formatter.cc#L1455) to `serializeAsProto` (the serializing function called when `TYPED` is used) returns `nullptr`. 

We could solve this problem by creating a new `F` option (i.e. `TLVS`, as I mentioned above) to explicitly format a TLV vector stored in the filter state. The benefit of this approach compared to the substitution formatter `PROXY_PROTOCOL_TLVS` I've already implemented is that we can pass in a different value for the filter state key used to access the TLV vector, but I'm not sure if that's desirable or not. Would love to hear your thoughts!

Reply:
That's because `serializeAsString` and `serializeAsProto` are virtual functions. You can override them in `ProxyProtocolFilterState`.

Reply:
@danhcole @gregbanks welcome to the thread we alluded to on slack! Would love to hear your thoughts on using the existing `FILTER_STATE` formatter v. creating a new `PROXY_PROTOCOL_TLVS` formatter.

Reply:
A benefit of creating a new formatter is that it'd be easy to only add TLVs of a specified type to the header by adding a param, i.e. `%PROXY_PROTOCOL_TLVS(<tlv_type>)%`, as was discussed on slack.

Reply:
@visualzhou oh 🤦‍♀️ that makes sense, and that makes using the existing `FILTER_STATE` formatter very straightforward

Reply:
Yeah, JSON might be too complex. I like the idea of specifying the key and just getting the value. That means SRE team needs to specify a header formatter for each cloud providers's private link TLV type. We could also go with whatever format works for Realm and leave the config and generic design to future work when merging this upstream later this year.

Reply:
@visualzhou that makes sense to me!

@danhcole @gregbanks how does this design sound to you? 

To be explicit, this formatter would be used by including the following in the envoy config:
```
            request_headers_to_add:
            - header:
                key: <desired_header_key>
                value: "%PROXY_PROTOCOL_TLVS(<tlv_type>)%"
```
and the following HTTP header would be added, where all included TLV values have type `<tlv_type>`:
```
<desired_header_key>: <base64_encoded_tlv_value_1>(, <base64_encoded_tlv_value_2>, ...)
```

Reply:
For our use case, `%PROXY_PROTOCOL_TLVS(<tlv_type)%"` works perfectly.  In terms of the feasibility of getting this upstreamed, if it's easier to do so using `%FILTER_STATE%` with a custom implantation of the TLVs that we can parse, I'd also be fine with that.

i.e. something like 
```yaml
            request_headers_to_add:
            - header:
                key: <desired_header_key>
                value: "%FILTER_STATE(<key>:TYPED)%"
```
which would result in
```
<desired_header_key>: {{"tlv_value_1": "base64_tlv_key_1"}, ...}
```

In this case, we can clearly document this format, and BaaS/whoever can parse the resulting JSON and get the TLV value they need.

Reply:
It seems other Proxy Protocol fields like downstream IP and port are in separate formatters, i.e. `%DOWNSTREAM_REMOTE_ADDRESS%`, so I think it would be simplest to go with `%PROXY_PROTOCOL_TLVS(<tlv_type>)%` as Betsy proposed for now. It's not clear to me which format will be easier to be accepted upstream now.

@danhcole do you want this feature soon using `%PROXY_PROTOCOL_TLVS(<tlv_type>)%` or wait for us to make a PR with upstream and get their approval on the format? This feature depends on merging the patch to pass TLV to upstream (someone from Google took it over), so it takes even longer but saves SRE and BAAS team time to potentially change it in the future. I'd prefer to deliver this earlier.

Reply:
That's a very good point, RE `%DOWNSTREAM_REMOTE_ADDRESS%` and other extracted values.  I'm very happy with `%PROXY_PROTOCOL_TLVS(<tlv_type>)%`, I think that's going to be easy for us to slot into our Envoy config.

Reply:
Sounds great, I'll push these changes soon!